File: dotnet/01-commands/tests/ComplexMethod/ComplexMethod/SearchInDictionary.cs
namespace ComplexMethod
{
    internal class SearchInDictionary
    {
        public bool Execute(Dictionary<string, int> dimensions, int valueToFind)
        {
            foreach (var item in dimensions)
            {
                if (item.Value == valueToFind)
                {
                    return VerifyIpAddress(item.Key);
                }
            }

            return false;
        }

        private bool VerifyIpAddress(string key)
        {
            if (key.Split('.').Length == 4 && (Blacklist.IsWhiteList(key) || !Blacklist.IsBlacklisted(key)))
            {
                return true;
            }
            else if (key.Split(':').Length == 8)
            {
                return true;
            }

            return false;
        }

        private bool VerifyIpAddressSimple(string key)
        {
            if (key.Split('.').Length == 4)
            {
                return true;
            }
            else if (key.Split(':').Length == 8)
            {
                return true;
            }

            return false;
        }
    }
}

File: nodejs/01-commands/tests/sourcecode/ComplexMethod/SearchInDictionary.js
const Blacklist = require('./Blacklist');

class SearchInDictionary {
    async execute(dimensions, valueToFind) {
        for (const key of Object.keys(dimensions)) {
            if (dimensions[key] === valueToFind) {
                return await this.verifyIpAddress(key);
            }
        }

        return false;
    }

    async verifyIpAddress(key) {
        try {
            if (key.split('.').length === 4 && 
                (await Blacklist.isWhiteList(key) || 
                !(await Blacklist.isBlacklisted(key)))) {
                return true;
            } else if (key.split(':').length === 8) {
                return true;
            }
        } catch (e) {
            console.error(e);
        }

        return false;
    }

    verifyIpAddressSimple(key) {
        if (key.split('.').length === 4) {
            return true;
        } else if (key.split(':').length === 8) {
            return true;
        }

        return false;
    }
}

module.exports = SearchInDictionary;